How to build an ACCESS model page (#821)
ACCESS-NRI/access-hive.org.au: docs/models/run-a-model/build_a_model.md

Author: This is the main content for your review.

@harshula @aidanheerdegen

Thanks you.

As an example, in the following instructions we will show how to build a modified [mom5 component](/models/model_components/ocean/#mom5) for [ACCESS-ESM1.5](/models/configurations/access-esm/#access-esm15). All other components and packages (i.e., dependencies) will remain the same as the official [ACCESS-ESM1.5 release]({{esm1_5_build_config}}).

!!! tip
    The instructions below remain valid (with simple tweaks) for any model or package.

If you are looking for information on how to run a model, refer to the [Run a Model](/models/run-a-model) section.

## Prerequisites

- **NCI account**<br>
  Before building a model, you need to [Set Up your NCI Account](/getting_started/set_up_nci_account).

- **_Spack_**<br>
  To set up _Spack_ on _Gadi_, refer to [Set up Spack for building ACCESS models]({{spack_setup}}).

## Navigate into your Spack directory
Navigate into the directory where you cloned the _Spack_ repositories during the [Spack setup]({{spack_setup}}#create-a-directory-for-spack).<br>
The suggested directory is `/g/data/$PROJECT/$USER/spack/0.22`:
```
cd /g/data/$PROJECT/$USER/spack/0.22
```

> ## Enable spack

Reviewer: Should we include the `module purge` during this step?

Author: Addressed in [cc82fe5632fba47c8b3c732f8fb9dbc6d190aedc](https://github.com/ACCESS-NRI/access-hive.org.au/pull/821/commits/cc82fe5632fba47c8b3c732f8fb9dbc6d190aedc)

!!! warning

>     This step needs to be carried out for any new login or new shell environment

Author: Addressed in [cc82fe5632fba47c8b3c732f8fb9dbc6d190aedc](https://github.com/ACCESS-NRI/access-hive.org.au/pull/821/commits/cc82fe5632fba47c8b3c732f8fb9dbc6d190aedc)

```
spack find mom5
```
<terminal-window>
  <terminal-line data="input" directory="[mom5_dev]">spack find mom5</terminal-line>
  <terminal-line>mom5@git.access-esm1.5_2024.08.23=access-esm1.5</terminal-line>
</terminal-window> 

To mark `mom5` as a development package and copy the `mom5@git.access-esm1.5_2024.08.23=access-esm1.5` version as a base copy, we can run:

```
spack develop mom5@git.access-esm1.5_2024.08.23=access-esm1.5
```
!!! tip
    This command should not display any output

This adds the following lines at the end of the `spack.yaml` file inside the [environment's folder](#spack-environment-folder):
```
develop:
    mom5:
      spec: mom5@git.access-esm1.5_2024.08.23=access-esm1.5
```
The source code is automatically copied inside the [environment's folder](#spack-environment-folder).

!!! warning
    As opposed to [Case 1](#case1), there is no `path` specification inside the `develop` portion of the `spack.yaml` environment configuration file.<br>
    This means that _Spack_ automatically expects to find the source code inside the [environment's folder](#spack-environment-folder).

## Compile modified Spack environment packages

> After the source code has been modified, the new package can be built following the same steps listed in [Compile Spack environment packages](#compile-spack-environment-packages-optional):
> 
> ```
> spack concretize -f --fresh
> spack install
> ```

Reviewer 2: ```suggestion
The package can be built following the same steps listed in [Compile Spack environment packages](#compile-spack-environment-packages-optional). Before making code changes to the package being developed, confirm that your environment can still be built and installed by running:

\```
spack concretize -f --fresh
spack install
\```

From now on, the _Spack_ environment only needs to be concretized when there are changes in the `spack.yaml` file.  You can modify the source code under development and directly install (by running `spack install` the environment without repeating the concretization step.
```

Author: I changed the structure a bit, mentioning that the concretization step only needs to be done if changes occur in the `config.yaml` file, and in this section I wrote the following:

---
## Compile modified Spack environment packages

After setting a development package the _Spack_ environment needs to be re-concretized (because the `spack.yaml` file changed). The _Spack_ environment can be concretized following the same steps listed in [Concretize the Spack environment](#concretize-the-spack-environment). Then, the new package can be built following the steps listed in [Compile Spack environment packages](#compile-spack-environment-packages-optional):

```
spack concretize -f --fresh
spack install
```
...
...

> [!TIP]
> From now on, the source code can be modified and the _Spack_ environment installed without repeating the concretization step.
> The _Spack_ environment will need to be re-concretized only if further changes occur in the `spack.yaml` file.

Author: Addressed in [cc82fe5632fba47c8b3c732f8fb9dbc6d190aedc](https://github.com/ACCESS-NRI/access-hive.org.au/pull/821/commits/cc82fe5632fba47c8b3c732f8fb9dbc6d190aedc)

  </terminal-line>
  <terminal-line lineDelay=200>
    <span class="green bold">[+]</span> /home/565/$USER/myspack/restricted/ukmo/release/linux-rocky8-x86_64_v4/intel-19.0.3.199/cice4-git.2024.05.21_access-esm1.5-hhtnigwxdyz7ta4dv3gvhwulze6hxqra
  </terminal-line>
  <terminal-line lineDelay=200>
    <span class="indigo bold">\==></span> <span class="highlighted bold">Installing</span> <span class="green bold">access-esm1p5-git.2024.05.1_2024.05.1-nkvasig2zrq2ocz6evva6bmurdq7nh3h</span> <span class="highlighted bold">[17/17]</span>
  </terminal-line>
  <terminal-line lineDelay=200>
    <span class="green bold">[+]</span> /home/565/$USER/myspack/restricted/ukmo/release/linux-rocky8-x86_64_v4/intel-19.0.3.199/access-esm1p5-git.2024.05.1_2024.05.1-nkvasig2zrq2ocz6evva6bmurdq7nh3h
  </terminal-line>
  <terminal-line lineDelay=200>
    <span class="indigo bold">\==></span> Updating view at /home/565/$USER/myspack/environments/mom5_dev/.spack-env/view
  </terminal-line>
</terminal-window>

!!! info
    The full output has been truncated for brevity.

## Output directory for compiled packages

> For the Spack instance obtained through the [Spack setup instructions]({{spack_setup}}), all compiled packages will be placed in directories having the following format: `<install_tree:root>/<architecture>/<compiler>/<package-name>`.

Reviewer: Could we please use the same directory structure naming that is in the config.yaml? i.e. `{architecture}/{compiler.name}-{compiler.version}/{name}-{version}-{hash}`

Author: Addressed in [cc82fe5632fba47c8b3c732f8fb9dbc6d190aedc](https://github.com/ACCESS-NRI/access-hive.org.au/pull/821/commits/cc82fe5632fba47c8b3c732f8fb9dbc6d190aedc)

!!! danger
    This page is tailored to experienced users and collaborators developing ACCESS models.

{% set esm1_5_build_config = "https://github.com/ACCESS-NRI/ACCESS-ESM1.5" %}
{% set spack_setup = "/getting_started/spack" %}
[gadi]: https://opus.nci.org.au/display/Help/0.+Welcome+to+Gadi#id-0.WelcometoGadi-Overview

# Build an ACCESS model

## About

The instructions below outline how to build an ACCESS model using the build-from-source package manager [Spack](https://spack.readthedocs.io).<br>
This build workflow is specifically designed to run on the [National Computating Infrastructure (NCI)](https://nci.org.au/about-us/who-we-are) supercomputer [_Gadi_][gadi].

> As an example, in the following instrutions we will show how to build a modified [mom5 component](/models/model_components/ocean/#mom5) for [ACCESS-ESM1.5](/models/configurations/access-esm/#access-esm15). All other components and packages (i.e., dependencies) will remain the same as the official [ACCESS-ESM1.5 release]({{esm1_5_build_config}}).

Reviewer 3: Typo: change "instrutions" to instructions

Author: Solved in commit [e47b759b407af36d1efb9212cf4a7fcc4a3dcdc8](https://github.com/ACCESS-NRI/access-hive.org.au/pull/821/commits/e47b759b407af36d1efb9212cf4a7fcc4a3dcdc8)

!!! tip
    The instructions below remain valid (with simple tweaks) for any model or package.

If you are looking for information on how to run a model, refer to the [Run a Model](/models/run-a-model) section.

## Prerequisites

- **NCI account**<br>
  Before building a model, you need to [Set Up your NCI Account](/getting_started/set_up_nci_account).

- **_Spack_**<br>
  To set up _Spack_ on _Gadi_, refer to [Set up Spack for building ACCESS models]({{spack_setup}}).

## Navigate into your Spack directory

> Navigate into the directory where you cloned the _Spack_ repositories during [Spack's setup]({{spack_setup}}#create-a-directory-for-spack).<br>

Reviewer 3: typo: change "during Spack's setup" to "during the Spack setup"

Author: Solved in commit [e47b759b407af36d1efb9212cf4a7fcc4a3dcdc8](https://github.com/ACCESS-NRI/access-hive.org.au/pull/821/commits/e47b759b407af36d1efb9212cf4a7fcc4a3dcdc8)

The suggested directory is `/g/data/$PROJECT/$USER/spack/0.22`:
```

> cd /g/data/$PROJECT/$USER/spack/0.22

Reviewer: Is it possible to avoid including a specific Spack version in this document? Could we use `$SPACKVERSION`?

Author: I understand your point. 

I think we are making concrete examples here and if we set `SPACKVERSION` to any version (`0.22` in this case) we would still have to update it when we update the major version.

I think we can easily (and quickly) replace all instances of `spack/0.22` with `spack/<new-version>` in the documentation once we have a new version released.
Also, for different models, users might want to use older `spack` versions (for back-compatibility) even though a newer version is released, so the problem would still persist also with something like `SPACKVERSION`. 

I think for now we should leave it like this, and focus the next steps on having a better Spack workflow (with `module load spack`) so many steps listed here (as well as the spack setup) is not needed at all.

```

## Enable spack

!!! warning

>     This step needs to be carried out for any new login or new shell

Reviewer 3: change "new shell" to "new shell environment."

Author: Solved in commit [e47b759b407af36d1efb9212cf4a7fcc4a3dcdc8](https://github.com/ACCESS-NRI/access-hive.org.au/pull/821/commits/e47b759b407af36d1efb9212cf4a7fcc4a3dcdc8)

To add the `spack` command to your shell, as well as other settings, run:
```
. spack-config/spack-enable.bash
```

!!! warning
    There is a space between the `.` and the path to the file, as we are [sourcing](https://tldp.org/HOWTO/Bash-Prompt-HOWTO/x237.html#:~:text=When%20a%20file%20is%20sourced,the%20file%20they%20are%20in.) the file.

## Create a Spack development environment

> _Spack_ has [environments](https://spack.readthedocs.io/en/latest/environments.html) that, in some ways, are similar to _Conda_ environments or _Python_ virtual environments: they create an isolated operating environment where _Spack_ can only see and access specific packages. This allows to build and deploy new packages in a coherent fashion.

Reviewer 3: change "that, in some ways, are similar" to "that are, in some ways, similar"

Author: Solved in commit [e47b759b407af36d1efb9212cf4a7fcc4a3dcdc8](https://github.com/ACCESS-NRI/access-hive.org.au/pull/821/commits/e47b759b407af36d1efb9212cf4a7fcc4a3dcdc8)

> _Spack_ environments are located inside `/g/data/$PROJECT/$USER/spack/0.22/environments`, each in a separate directory. Each environment configuration is defined within its directory using the [`spack.yaml`](https://spack.readthedocs.io/en/latest/config_yaml.html#config-yaml) file.

Reviewer: ```suggestion
_Spack_ `managed environments` are located inside `/g/data/$PROJECT/$USER/spack/0.22/environments`, each in a separate directory. Each environment configuration is defined within its directory using the [`spack.yaml`](https://spack.readthedocs.io/en/latest/config_yaml.html#config-yaml) file.
```
```suggestion
_Spack_ environments are located inside `/g/data/$PROJECT/$USER/spack/0.22/environments`, each in a separate directory. Each environment configuration is defined within its directory using the [`spack.yaml`](https://spack.readthedocs.io/en/latest/config_yaml.html#config-yaml) file.
```
```suggestion
_Spack_ environments are located inside `/g/data/$PROJECT/$USER/spack/0.22/environments`, each in a separate directory. Each environment configuration is defined within its directory using the [`spack.yaml`](https://spack.readthedocs.io/en/latest/config_yaml.html#config-yaml) file.
```

Author: Addressed in [13506e3c5e8c940d2ffd71ea48a334ce4f1046f4](https://github.com/ACCESS-NRI/access-hive.org.au/pull/821/commits/13506e3c5e8c940d2ffd71ea48a334ce4f1046f4).

{: id="spack-environment-folder"}

!!! warning

>     _Spack_ environments' location can be changed within _Spack_ configuration files and the directory specified above represents the default location for a _Spack_ instance that has been set up following [Spack's setup instructions]({{spack_setup}}).

Reviewer 3: change: "following [Spack's setup instructions]..." to "following the [Spack setup] instructions..."

Author: Solved in commit [e47b759b407af36d1efb9212cf4a7fcc4a3dcdc8](https://github.com/ACCESS-NRI/access-hive.org.au/pull/821/commits/e47b759b407af36d1efb9212cf4a7fcc4a3dcdc8)

  <terminal-line>Cloning into 'ACCESS-ESM1.5'...</terminal-line>
  <terminal-line>remote: Enumerating objects: 29, done.</terminal-line>
  <terminal-line>remote: Counting objects: 100% (19/19), done.</terminal-line>
  <terminal-line>remote: Compressing objects: 100% (17/17), done.</terminal-line>
  <terminal-line>remote: Total 29 (delta 6), reused 3 (delta 1), pack-reused 10 (from 1)</terminal-line>
  <terminal-line>Receiving objects: 100% (29/29), 16.40 KiB | 2.38 MiB/s, done.</terminal-line>
  <terminal-line>Resolving deltas: 100% (6/6), done.</terminal-line>
  <terminal-line data="input">spack env create mom5_dev</terminal-line>
  <terminal-line><span class="indigo bold">\==></span> Created environment <span class="cyan">mom5_dev</span> in: <span class="cyan">/g/data/$PROJECT/$USER/spack/0.22/environments/mom5_dev</span></terminal-line>
  <terminal-line><span class="indigo bold">\==></span> Activate with: <span class="cyan">spack env activate mom5_dev</span></terminal-line>
</terminal-window>

The newly-created `mom5_dev` _Spack_ <b>environment folder</b> is `/g/data/$PROJECT/$USER/spack/0.22/environments/mom5_dev`.

### Activate the environment

> To activate the `mom5_dev` _Spack_ environment run:

Reviewer 3: Add a comma after environment to be consistent with previous use. i.e., "...environment, run:"

Author: Solved in commit [e47b759b407af36d1efb9212cf4a7fcc4a3dcdc8](https://github.com/ACCESS-NRI/access-hive.org.au/pull/821/commits/e47b759b407af36d1efb9212cf4a7fcc4a3dcdc8)

```
spack env activate -p mom5_dev
```
<terminal-window>
    <terminal-line data="input">spack env activate -p mom5_dev</terminal-line>
    <terminal-line data="input" directory="[mom5_dev]" lineDelay=0></terminal-line>
</terminal-window>

## Compile Spack environment packages (optional)

> It is always good to first compile all the packages in the newly created _Spack_ environment as is (without changing).

Reviewer 3: Perhaps change "It is always good to..." to "It is recommended to..." and 
"as is (without changing)." to "without making any changes."

Author: Solved in commit [e47b759b407af36d1efb9212cf4a7fcc4a3dcdc8](https://github.com/ACCESS-NRI/access-hive.org.au/pull/821/commits/e47b759b407af36d1efb9212cf4a7fcc4a3dcdc8)

Compiling all the packages present in a _Spack_ environment is referred to as [installing the environment](https://spack.readthedocs.io/en/latest/environments.html#installing-an-environment).

### Concretize the Spack environment
[Concretizing](https://spack.readthedocs.io/en/latest/environments.html#concretizing) the _Spack_ environment is necessary to force _Spack_ to update its knowledge of all the packages (their specs) within the environment.

> To concretize the `mom5_dev` environment run:

Reviewer 3: As before, add a comma after environment to be consistent with previous use. i.e., "...environment, run:"

Author: Solved in commit [e47b759b407af36d1efb9212cf4a7fcc4a3dcdc8](https://github.com/ACCESS-NRI/access-hive.org.au/pull/821/commits/e47b759b407af36d1efb9212cf4a7fcc4a3dcdc8)

  </terminal-line>
  <terminal-line>
    <span class="grey preserve"> -   qy5w2d7        </span> ^dummygrib<span class="cyan">@1.0</span><span class="green">%intel@19.0.3.199</span> <span class="indigo">build_system=makefile</span> <span class="pink">arch=linux-rocky8-x86_64_v4</span>
  </terminal-line>
  <terminal-line>
    <span class="grey preserve"> -   ho2ie66        </span> ^fcm<span class="cyan">@2021.05.0</span><span class="green">%intel@19.0.3.199</span> <span class="indigo">build_system=generic site=none</span> <span class="pink">arch=linux-rocky8-x86_64_v4</span>
  </terminal-line>
  <terminal-line>
    <span class="grey preserve"> -   xalavwv        </span> ^gcom4<span class="cyan">@git.2024.05.28=access-esm1.5</span><span class="green">%intel@19.0.3.199</span><span class="indigo">+mpi build_system=generic</span> <span class="pink">arch=linux-rocky8-x86_64_v4</span>
  </terminal-line>
</terminal-window>

### Install the Spack environment

> To compile the packages in the `mom5_dev` environment run:

Reviewer 3: As before, add a comma after environment to be consistent with previous use. i.e., "...environment, run:"

Author: Solved in commit [e47b759b407af36d1efb9212cf4a7fcc4a3dcdc8](https://github.com/ACCESS-NRI/access-hive.org.au/pull/821/commits/e47b759b407af36d1efb9212cf4a7fcc4a3dcdc8)

```
spack install 
```

!!! tip

>     It takes 30-40 mins to compile all the packages for the first time.<br>

Reviewer 3: change "mins" to "minutes"

Author: Solved in commit [e47b759b407af36d1efb9212cf4a7fcc4a3dcdc8](https://github.com/ACCESS-NRI/access-hive.org.au/pull/821/commits/e47b759b407af36d1efb9212cf4a7fcc4a3dcdc8)

>     However, subsequent installations will re-use the built dependencies, therefore the total compilation time will be reduced.

Reviewer 3: Perhaps reword sentence: "Subsequent installations, however, will compile quicker as the built dependencies are reused."

Author: Solved in commit [e47b759b407af36d1efb9212cf4a7fcc4a3dcdc8](https://github.com/ACCESS-NRI/access-hive.org.au/pull/821/commits/e47b759b407af36d1efb9212cf4a7fcc4a3dcdc8)

  <terminal-line>
    <span class="indigo bold">\==></span> <span class="highlighted bold">Installing</span> <span class="green bold">cice4-git.2024.05.21=access-esm1.5-hhtnigwxdyz7ta4dv3gvhwulze6hxqra</span> <span class="highlighted bold">[16/17]</span>
  </terminal-line>
  <terminal-line>
    <span class="green bold">[+]</span> /home/565/$USER/myspack/restricted/ukmo/release/linux-rocky8-x86_64_v4/intel-19.0.3.199/cice4-git.2024.05.21_access-esm1.5-hhtnigwxdyz7ta4dv3gvhwulze6hxqra
  </terminal-line>
  <terminal-line>
    <span class="indigo bold">\==></span> <span class="highlighted bold">Installing</span> <span class="green bold">access-esm1p5-git.2024.05.1=2024.05.1-nkvasig2zrq2ocz6evva6bmurdq7nh3h</span> <span class="highlighted bold">[17/17]</span>
  </terminal-line>
  <terminal-line>
    <span class="green bold">[+]</span> /home/565/$USER/myspack/restricted/ukmo/release/linux-rocky8-x86_64_v4/intel-19.0.3.199/access-esm1p5-git.2024.05.1_2024.05.1-nkvasig2zrq2ocz6evva6bmurdq7nh3h
  </terminal-line>
  <terminal-line>
    <span class="indigo bold">\==></span> Updating view at /home/565/$USER/myspack/environments/mom5_dev/.spack-env/view
  </terminal-line>
</terminal-window>

> !!! info

Reviewer 2: There should be a section here on changing the versions of packages, without using spack develop.

i.e. how do i us a component @git.main / @git.commit or a fork and build from there ?

Author: Hi @anton-seaice, 

Thank you for the comment. 

If I understand correctly what you mean, you would have to clone the repo on Gadi and use the instructions listed in [case1](https://access-hive.org.au/pr-preview/pr-821/models/run-a-model/build_a_model/#case1) (`spack develop mom5@label --path /path/to/your/cloned/repo`).

Did you want to do anything different?

Reviewer 2: I was thinking of the case where you just change which version of a component is built.

i.e. in this examples in this doc, the environment file requests:

`cice4@git.2024.05.21`

but instead i want to build from a branch:

`cice4@git.main`

or a commit :

`cice4@git.6d8b9d1ee7e4a07cb4cebeb4dcf00592e961519b`

Which should be able to be done without using spack develop, (noting spack develop looses provenance of the build).

Reviewer 4: This is one the identified use-cases that wasn't covered in the initial developer docs instructions, but was deferred to a later date as it was clear it would need careful testing and documentation.

We do want to do this at some stage, just not sure we necessarily want to do so right now. But I'm happy to acquiesce if there is a consensus we should do this now.

Reviewer 2: I think it's a very common case / thing to want to do :)

Author: Merging as it is for now.
@anton-seaice please feel free to raise a separate issue for your workflow.

    The full output has been truncated for brevity.

## Create development package

When you develop a package within a _Spack_ environment, _Spack_ needs to know that the desired package is marked as "in development", and be able to access its source code.<br>
This is done through the [`spack develop`](https://spack.readthedocs.io/en/latest/command_index.html#spack-develop) command.

### Mark package as a development package

> There are in general two cases that influence the command to run to mark a package as "in development", depending on the state of the package's source code:

Reviewer 3: Edit: "There are, in general, two cases that influence the run command to mark..."

Author: Solved in commit [e47b759b407af36d1efb9212cf4a7fcc4a3dcdc8](https://github.com/ACCESS-NRI/access-hive.org.au/pull/821/commits/e47b759b407af36d1efb9212cf4a7fcc4a3dcdc8)

1. [The new source code already exists](#case1)

> 2. [The new source code does not exist yet](#case2)

Reviewer 3: edit: "does not yet exist"

Author: Solved in commit [e47b759b407af36d1efb9212cf4a7fcc4a3dcdc8](https://github.com/ACCESS-NRI/access-hive.org.au/pull/821/commits/e47b759b407af36d1efb9212cf4a7fcc4a3dcdc8)

   
#### Case 1. The new source code already exists {: id='case1'}
If the new source code for the development package already exists in the filesystem (this also includes cases when the new source code is in a `git` repo that can be cloned), to mark a package as "in development" we will need two things:

> - A name for the package new tag

Reviewer: ```suggestion
- A name for the package's new tag
```

Author: Addressed in [13506e3c5e8c940d2ffd71ea48a334ce4f1046f4](https://github.com/ACCESS-NRI/access-hive.org.au/pull/821/commits/13506e3c5e8c940d2ffd71ea48a334ce4f1046f4).

- The path of the new source code

> For example, to mark the `mom5` package as a development package using the `development_version` tag and with the new source code in the `/path/to/mom5/new/source/code` folder, we can run:

Reviewer: ```suggestion
For example, to mark `mom5` package as a development package using the `development_version` tag and with the new source code in the `/path/to/mom5/new/source/code` folder, we can run:
```

Author: Addressed in [13506e3c5e8c940d2ffd71ea48a334ce4f1046f4](https://github.com/ACCESS-NRI/access-hive.org.au/pull/821/commits/13506e3c5e8c940d2ffd71ea48a334ce4f1046f4).

In this case we need to tell _Spack_ which package tag (version) to copy as a starting base for the new source code.

For example, to retrieve the tag for the original `mom5` package used in the `mom5_dev` _Spack_ environment, we can run:
```
spack find mom5
```
<terminal-window>
  <terminal-line data="input" directory="[mom5_dev]">spack find mom5</terminal-line>
  <terminal-line data="input">mom5@git.access-esm1.5_2024.08.23=access-esm1.5</terminal-line>
</terminal-window> 

To mark the `mom5` as a development package and copy the `mom5@git.access-esm1.5_2024.08.23=access-esm1.5` version as a base copy, we can run:

```
spack develop mom5@git.access-esm1.5_2024.08.23=access-esm1.5
```
!!! tip
    This command should not display any output

This adds the following lines at the end of the `spack.yaml` file inside the [environment's folder](#spack-environment-folder):
```
develop:
    mom5:
      spec: mom5@git.access-esm1.5_2024.08.23=access-esm1.5
```
The source code is automatically copied inside the [environment's folder](#spack-environment-folder).

!!! warning

>     As you might notice, differently from [Case 1](#case1), there is no `path` specification inside the `develop` portion of the `spack.yaml` environment configuration file.<br>

Reviewer 3: Edit: "As opposed to [Case 1](#case1), there is no `path` specification inside..."

Author: Solved in commit [e47b759b407af36d1efb9212cf4a7fcc4a3dcdc8](https://github.com/ACCESS-NRI/access-hive.org.au/pull/821/commits/e47b759b407af36d1efb9212cf4a7fcc4a3dcdc8)

    This means that _Spack_ automatically expects to find the source code inside the [environment's folder](#spack-environment-folder).

## Compile modified Spack environment packages

After the source code has been modified, the new package can be built following the same steps listed in [Compile Spack environment packages](#compile-spack-environment-packages-optional):

```
spack concretize -f --fresh
spack install

> ```

Reviewer 2: I think after you've done the spack concretize, you can run spack install as many times as needed right ?

i.e. make another change a change to the source code being developed, then spack install

and you don't need to concretize again

Author: I think you would need to concretize only after you make changes in the environment configuration / settings (for example after you set a development package), to update spack "knowledge" about those changes. 
You wouldn't need to concretize again after changing the source code being developed. 

Am I correct @harshula?
If that's the case, it might be good to add an info box that specifies it.

Reviewer: My understanding is that `spack concretize` is needed when there's a change to the `spack.yaml`.

Reviewer 2: The document says this :

> After the source code has been modified, the new package can be built following the same steps listed in [Compile Spack environment packages](https://access-hive.org.au/pr-preview/pr-821/models/run-a-model/build_a_model/#compile-spack-environment-packages-optional):
> 
> spack concretize -f --fresh ...
> spack install

But this is not quite true - You only need to spack concretize after changing the spack.yaml

When modifying the source code, you only need to spack install

Author: You are right, but in that specific case we modified the `spack.yaml` (the step right before the part you referenced contains the `spack develop` command that modified the `spack.yaml` file).

It is true that, in genera, you don't need to re-concretize the environment every time you modify your code if you already concretized after running `spack develop`. 
However, I don't see any harm in concretizing again. 
Therefore, for simplicity, I would leave the documentation as it is. 

What do you think?

Reviewer 2: Its likely that mostly you would only do spack develop & spack concretise once, but then change the source code many times and only do a spack install for each change

Spack concretise take 5+ minutes ? Which is pretty annoying if you are trying to just find bugs in your code or whatever.

Author: Yes that is true. 
Ok I will add a `tip` admonition that says:

> [!TIP]
> The _Spack_ environment needs to be concretized only when there are changes in the `spack.yaml` file.
> Therefore, after the first concretization after setting the development package, if any modification of the source code occurs, you can install the environment directly by skipping the concretization step.

Does it sound good? 
Please feel free to modify the text I wrote above however you think would make more sense.

Author: Improved version (thanksk to ChatGPT):

> [!TIP]
> The _Spack_ environment only needs to be concretized when there are changes in the `spack.yaml` file. After the initial concretization following the setup of the development package, you can directly install the environment without repeating the concretization step, even if source code modifications occur.

Reviewer 2: I added a suggestion - see https://github.com/ACCESS-NRI/access-hive.org.au/pull/821/files#r1796194338

Author: Addressed in [cc82fe5632fba47c8b3c732f8fb9dbc6d190aedc](https://github.com/ACCESS-NRI/access-hive.org.au/pull/821/commits/cc82fe5632fba47c8b3c732f8fb9dbc6d190aedc)

!!! warning

>     Even though the `spack install` command this time will only build the development package, it might still take a long time to complete, depending on the specific package.

Reviewer 3: Edit: "Although this time the `spack install` command will only build the development package, it still may take a while to complete, depending on the specific package."

Author: Solved in commit [e47b759b407af36d1efb9212cf4a7fcc4a3dcdc8](https://github.com/ACCESS-NRI/access-hive.org.au/pull/821/commits/e47b759b407af36d1efb9212cf4a7fcc4a3dcdc8)

  </terminal-line>
  <terminal-line lineDelay=200>
    <span class="green bold">[+]</span> /home/565/$USER/myspack/restricted/ukmo/release/linux-rocky8-x86_64_v4/intel-19.0.3.199/access-esm1p5-git.2024.05.1_2024.05.1-nkvasig2zrq2ocz6evva6bmurdq7nh3h
  </terminal-line>
  <terminal-line lineDelay=200>
    <span class="indigo bold">\==></span> Updating view at /home/565/$USER/myspack/environments/mom5_dev/.spack-env/view
  </terminal-line>
</terminal-window>

!!! info
    The full output has been truncated for brevity.

## Output directory for compiled packages
All compiled packages will be placed in directories having the following format: `<install_tree:root>/<architecture>/<compiler>/<package-name>`.

`<install_tree:root>` depends on the [`install_tree`](https://spack.readthedocs.io/en/latest/config_yaml.html#install-tree-root) configuration field. _Spack_ reads this configuration field from files in several directories, following [Spack's configuration scopes](https://spack.readthedocs.io/en/latest/configuration.html#configuration-scopes).

!!! warning
    For instances of _Spack_ on _Gadi_ you should ignore the **system** scope.

!!! tip

>     For the _Spack_ instance obtained through the [Spack's setup instructions]({{spack_setup}}), `$(prefix)` corresponds to the `/g/data/$PROJECT/$USER/spack/0.22/spack` directory.

Reviewer 3: change: "the [Spack's setup instructions]..." to "the [Spack setup] instructions..."

Author: Solved in commit [e47b759b407af36d1efb9212cf4a7fcc4a3dcdc8](https://github.com/ACCESS-NRI/access-hive.org.au/pull/821/commits/e47b759b407af36d1efb9212cf4a7fcc4a3dcdc8)

For the example above, `mom5_dev` _Spack_ environment's configuration file (`spack.yaml`) contains the following lines that fall in the **environment** scope:
```yaml
config:
    install_tree:
      root: $spack/../restricted/ukmo/release
```
!!! tip
    `$spack` refers to the `$(prefix)` directory above.

> This means the packages that we just built in this example can be found under `/g/data/$PROJECT/$USER/spack/0.22/spack/../restricted/ukmo/release/<architecture>/<compiler>/<package-name>`.

Reviewer 3: Edit: "This means the packages built in this example can be found in... "

Author: Solved in commit [e47b759b407af36d1efb9212cf4a7fcc4a3dcdc8](https://github.com/ACCESS-NRI/access-hive.org.au/pull/821/commits/e47b759b407af36d1efb9212cf4a7fcc4a3dcdc8)

## Troubleshooting build errors

> Sometimes you might get errors while compiling the packages.<br>

Reviewer 3: replace "get" with "encounter"

Author: Solved in commit [e47b759b407af36d1efb9212cf4a7fcc4a3dcdc8](https://github.com/ACCESS-NRI/access-hive.org.au/pull/821/commits/e47b759b407af36d1efb9212cf4a7fcc4a3dcdc8)

<pre><code><spack class="grey bold">...</spack>
<span class="red"> >> 415    /home/565/dm5220/myspack/environments/mom5_dev/mom5/src/access_coupler/ocean_solo.F90:224: undefined reference to `constant s_init_'</span>
<span class="red"> >> 416    make: *** [Makefile:931: fms_ACCESS-CM.x] Error 1</span>
<spack class="grey bold">...</spack>
See build log for details:
  /scratch/$PROJECT/$USER/tmp/path/to/the/spack-stage-mom5-git.access-esm1.5_2024.08.23_access-esm1.5-l34w7is54xzer7s4ztvb5ymgjbtduknh/spack-build-out.txt
<spack class="grey bold">...</spack>
<span class="red">==></span> Error: access-esm1p5-git.2024.05.1=2024.05.1-aysea5r7rbwy22lluvl64baperlokktv: Package was not installed
<span class="red">==></span> Error: Installation request failed.  Refer to reported errors for failing package(s).</code></pre>

If the error is not obvious from the error message, see the build log for more information.

> 

Reviewer 2: I think at some point we need a section on how to share the resulting modified code? Maybe thats a seperate doc?

But it would cover commiting the new-code somewhere, and modifying the environment to build from the git source ? and then committing the environment somewhere it can be shared ?

Reviewer: My understanding is that we plan to encourage sharing of modified code via PRs and Pre-release builds. It would make sense to have a *separate* document with instructions on how to do that.

> In this case the error might be obvious already from the error message but, for more information, the build log can be examined.

Reviewer 3: Edit: "If the error is not obvious from the error message, see the build log for more information."

Author: Solved in commit [e47b759b407af36d1efb9212cf4a7fcc4a3dcdc8](https://github.com/ACCESS-NRI/access-hive.org.au/pull/821/commits/e47b759b407af36d1efb9212cf4a7fcc4a3dcdc8)

  <terminal-line>Receiving objects: 100% (29/29), 16.40 KiB | 2.38 MiB/s, done.</terminal-line>
  <terminal-line>Resolving deltas: 100% (6/6), done.</terminal-line>
  <terminal-line data="input">spack env create mom5_dev</terminal-line>
  <terminal-line><span class="indigo bold">\==></span> Created environment <span class="cyan">mom5_dev</span> in: <span class="cyan">/g/data/$PROJECT/$USER/spack/0.22/environments/mom5_dev</span></terminal-line>
  <terminal-line><span class="indigo bold">\==></span> Activate with: <span class="cyan">spack env activate mom5_dev</span></terminal-line>
</terminal-window>

The newly-created `mom5_dev` _Spack_ <b>environment folder</b> is `/g/data/$PROJECT/$USER/spack/0.22/environments/mom5_dev`.

### Activate the environment
To activate the `mom5_dev` _Spack_ environment, run:
```
spack env activate -p mom5_dev
```
<terminal-window>
    <terminal-line data="input">spack env activate -p mom5_dev</terminal-line>
    <terminal-line data="input" directory="[mom5_dev]" lineDelay=0></terminal-line>
</terminal-window>

## Compile Spack environment packages (optional)

> It is recommended to first compile all the packages in the newly created _Spack_ environment as is without making any changes.

Reviewer: ```suggestion
It is recommended to first compile all the packages in the newly created _Spack_ environment, as is, without making any changes.
```

Author: Addressed in [260fa6d66defdca7bf12c52bbfcfc9a3e80d34c7](https://github.com/ACCESS-NRI/access-hive.org.au/pull/821/commits/260fa6d66defdca7bf12c52bbfcfc9a3e80d34c7)

  </terminal-line>
  <terminal-line lineDelay=200>
    <span class="green bold">[+]</span> /home/565/$USER/myspack/restricted/ukmo/release/linux-rocky8-x86_64_v4/intel-19.0.3.199/cice4-git.2024.05.21_access-esm1.5-hhtnigwxdyz7ta4dv3gvhwulze6hxqra
  </terminal-line>
  <terminal-line lineDelay=200>
    <span class="indigo bold">\==></span> <span class="highlighted bold">Installing</span> <span class="green bold">access-esm1p5-git.2024.05.1_2024.05.1-nkvasig2zrq2ocz6evva6bmurdq7nh3h</span> <span class="highlighted bold">[17/17]</span>
  </terminal-line>
  <terminal-line lineDelay=200>
    <span class="green bold">[+]</span> /home/565/$USER/myspack/restricted/ukmo/release/linux-rocky8-x86_64_v4/intel-19.0.3.199/access-esm1p5-git.2024.05.1_2024.05.1-nkvasig2zrq2ocz6evva6bmurdq7nh3h
  </terminal-line>
  <terminal-line lineDelay=200>
    <span class="indigo bold">\==></span> Updating view at /home/565/$USER/myspack/environments/mom5_dev/.spack-env/view
  </terminal-line>
</terminal-window>

!!! info
    The full output has been truncated for brevity.

!!! tip

>   From now on, the source code can be modified and the _Spack_ environment installed without repeating the concretization step.<br>

Reviewer: ```suggestion
    From now on, the source code can be modified and the _Spack_ environment installed without repeating the concretization step.<br>
```

Author: Addressed in [260fa6d66defdca7bf12c52bbfcfc9a3e80d34c7](https://github.com/ACCESS-NRI/access-hive.org.au/pull/821/commits/260fa6d66defdca7bf12c52bbfcfc9a3e80d34c7)

  The _Spack_ environment will need to be re-concretized only if further changes occur in the `spack.yaml` file.

## Output directory for compiled packages

> For the Spack instance obtained through the [Spack setup instructions]({{spack_setup}}), all compiled packages will be placed in directories having the following format: `<install_tree.root>/{architecture}/{compiler.name}-{compiler.version}/{name}-{version}-{hash}`.

Reviewer: ```suggestion
For the Spack instance obtained through the [Spack setup instructions]({{spack_setup}}), all compiled packages will be placed in directories having the following format: `<install_tree.root>/<architecture>/<compiler.name>-<compiler.version>/<name>-<version>-<hash>`.
```

Author: Addressed in [260fa6d66defdca7bf12c52bbfcfc9a3e80d34c7](https://github.com/ACCESS-NRI/access-hive.org.au/pull/821/commits/260fa6d66defdca7bf12c52bbfcfc9a3e80d34c7)

`<install_tree.root>` depends on the [`install_tree.root`](https://spack.readthedocs.io/en/latest/config_yaml.html#install-tree-root) configuration field. _Spack_ reads this configuration field from files in several directories, following [Spack's configuration scopes](https://spack.readthedocs.io/en/latest/configuration.html#configuration-scopes).

!!! warning
    For instances of _Spack_ on _Gadi_ you should ignore the **system** scope.

For the example above, `mom5_dev` _Spack_ environment's configuration file (`spack.yaml`) contains the following lines that fall in the **environment** scope:
```yaml
config:
    install_tree:
      root: $spack/../restricted/ukmo/release
```
!!! tip
    For the _Spack_ instance obtained through the [Spack setup instructions]({{spack_setup}}), `$spack` (referred to as `$(prefix)` in _Spack_ configuration scopes documentation above) corresponds to the `/g/data/$PROJECT/$USER/spack/0.22/spack` directory.

> This means the packages built in this example can be found in `/g/data/$PROJECT/$USER/spack/0.22/spack/../restricted/ukmo/release/<architecture>/<compiler>/<package-name>`.

Reviewer: ```suggestion
This means the packages built in this example can be found in `/g/data/$PROJECT/$USER/spack/0.22/spack/../restricted/ukmo/release/<architecture>/<compiler.name>-<compiler.version>/<name>-<version>-<hash>`.
```

Author: Addressed in [260fa6d66defdca7bf12c52bbfcfc9a3e80d34c7](https://github.com/ACCESS-NRI/access-hive.org.au/pull/821/commits/260fa6d66defdca7bf12c52bbfcfc9a3e80d34c7)
